File: extract_ismrmrd_parameters_from_headers.m
function h = extract_ismrmrd_parameters_from_headers(head, hdr, table_position_offset)
%create_nifti_parameters : function that returns a structure h which
%contains parameters needed for conversion function

%% Get index of the first real data
m=mean(head.position,1);
index=find(m~=0);
first_index_ok=index(1);
last_index_ok = index(end);


%% Create structure h
h = struct('NumberOfTemporalPositions',1);

%fill it with appropriate values
h.MRAcquisitionType = '3D';

% obscur (why a substraction ??)
h.ImageOrientationPatient = [ head.phase_dir(:,first_index_ok)' -head.read_dir(:,first_index_ok)']

h.SpacingBetweenSlices = hdr.encoding.reconSpace.fieldOfView_mm.z/hdr.encoding.reconSpace.matrixSize.z;

if (hdr.encoding.reconSpace.matrixSize.z == 0)
    hdr.encoding.reconSpace.matrixSize.z = 1;
end

h.SliceThickness = hdr.encoding.reconSpace.fieldOfView_mm.z/hdr.encoding.reconSpace.matrixSize.z;

h.PixelSpacing = [hdr.encoding.reconSpace.fieldOfView_mm.x / hdr.encoding.reconSpace.matrixSize.x ...
                  hdr.encoding.reconSpace.fieldOfView_mm.y / hdr.encoding.reconSpace.matrixSize.y ]';
              
% % ImagePositionPatient of first slice 
% % Attention peut-être que faut changer les +/-
% % en fonction de HFS dans : hdr.measurementInformation.patientPosition

% Creates a directions matrix and print it ((
directions = [head.read_dir(1,first_index_ok)  head.read_dir(2,first_index_ok)  head.read_dir(3,first_index_ok);
              head.phase_dir(1,first_index_ok) head.phase_dir(2,first_index_ok) head.phase_dir(3,first_index_ok);
              head.slice_dir(1,first_index_ok) head.slice_dir(2,first_index_ok) head.slice_dir(3,first_index_ok)]'
          


%Compute the corner position (ImagePositionPatient) of the first slice
corner_first_slice(1)= head.position(1)  ... 
    + (hdr.encoding.reconSpace.fieldOfView_mm.x / 2.0 ) * head.read_dir(1,first_index_ok) ...
    - (hdr.encoding.reconSpace.fieldOfView_mm.y / 2.0 ) * head.phase_dir(1,first_index_ok) ...
    - (hdr.encoding.reconSpace.fieldOfView_mm.z / 2.0 ) * head.slice_dir(1,first_index_ok) ;

corner_first_slice(2)=head.position(2)   ... 
    + (hdr.encoding.reconSpace.fieldOfView_mm.x / 2.0) * head.read_dir(2,first_index_ok) ...
    - (hdr.encoding.reconSpace.fieldOfView_mm.y / 2.0 ) * head.phase_dir(2,first_index_ok) ...
    - (hdr.encoding.reconSpace.fieldOfView_mm.z / 2.0) * head.slice_dir(2,first_index_ok) ;
   
corner_first_slice(3)=head.position(3) + table_position_offset  ...
    + (hdr.encoding.reconSpace.fieldOfView_mm.x / 2.0 ) * head.read_dir(3,first_index_ok) ...
    - (hdr.encoding.reconSpace.fieldOfView_mm.y / 2.0 ) * head.phase_dir(3,first_index_ok) ...
    - (hdr.encoding.reconSpace.fieldOfView_mm.z / 2.0 ) * head.slice_dir(3,first_index_ok) ;

h.ImagePositionPatient = corner_first_slice';

%Compute the corner position (ImagePositionPatient) of the last slice
corner_last_slice(1)=head.position(1)...
    + (hdr.encoding.reconSpace.fieldOfView_mm.x / 2.0 ) * head.read_dir(1,last_index_ok) ...
    - (hdr.encoding.reconSpace.fieldOfView_mm.y / 2.0 ) * head.phase_dir(1,last_index_ok) ...
    + (hdr.encoding.reconSpace.fieldOfView_mm.z / 2.0 ) * head.slice_dir(1,last_index_ok) ;

corner_last_slice(2)=head.position(2) ... 
    + (hdr.encoding.reconSpace.fieldOfView_mm.x / 2.0 ) * head.read_dir(2,last_index_ok) ...
    - (hdr.encoding.reconSpace.fieldOfView_mm.y / 2.0 ) * head.phase_dir(2,last_index_ok) ...
    + (hdr.encoding.reconSpace.fieldOfView_mm.z / 2.0 ) * head.slice_dir(2,last_index_ok) ;

corner_last_slice(3)=head.position(3) + table_position_offset ...
    + (hdr.encoding.reconSpace.fieldOfView_mm.x / 2.0 ) * head.read_dir(3,last_index_ok) ...
    - (hdr.encoding.reconSpace.fieldOfView_mm.y / 2.0 ) * head.phase_dir(3,last_index_ok) ...
    + (hdr.encoding.reconSpace.fieldOfView_mm.z / 2.0 ) * head.slice_dir(3,last_index_ok) ;

h.LastFile.ImagePositionPatient = corner_last_slice';

h.Manufacturer=hdr.acquisitionSystemInformation.systemVendor;

h.Columns = hdr.encoding.reconSpace.matrixSize.y;

h.NiftiName=hdr.measurementInformation.protocolName;

h.InPlanePhaseEncodingDirection='COL'; %or 'ROW'


end

File: python_version/extract_ismrmrd_parameters_from_headers.py
import numpy as np
import sys

def extract_ismrmrd_parameters_from_headers(head, hdr):
    """

    Parameters
    ----------
    head : read imsrmrd dataset head (dataset.head)
    hdr : deserialized ismrmrd xml dataset file
    Returns
    -------
    function that returns a structure h which contains parameters needed for conversion function
    """

    #get encoding parameters from hdr file
    enc = hdr.encoding[0]
    # Matrix size
    rNx = enc.reconSpace.matrixSize.x
    rNy = enc.reconSpace.matrixSize.y
    rNz = enc.reconSpace.matrixSize.z
    # Field of View
    rFOVx = enc.reconSpace.fieldOfView_mm.x
    rFOVy = enc.reconSpace.fieldOfView_mm.y
    rFOVz = enc.reconSpace.fieldOfView_mm.z

    # Create structure h and fill it with appropriate values
    h = {"NumberOfTemporalPositions": 1,
         "MRAcquisitionType": "3D",
         "ImageOrientationPatient": np.concatenate((-np.array(head.phase_dir), -np.array(head.read_dir)), axis=0),
         "SpacingBetweenSlices": rFOVz / rNz,
         "SliceThickness": rFOVz / rNz,
         "PixelSpacing": np.array([[rFOVx / rNx], [rFOVy / rNy]])}

    if (hdr.encoding[0].reconSpace.matrixSize.z == 0):
        hdr.encoding[0].reconSpace.matrixSize.z = 1


    #ImagePositionPatient of first slice (right corner position of first and last slice)
    #Attention peut-être que faut changer les +/-
    #en fonction de HFS dans : hdr.measurementInformation.patientPosition
    corner_first_slice = np.zeros((3, 1), dtype=float)
    corner_last_slice = np.zeros((3, 1), dtype=float)

    #Compute the corner position (ImagePositionPatient) of the first slice
    corner_first_slice[0] = head.position[0] \
                            + rFOVx / 2 * head.read_dir[0] \
                            + rFOVy / 2 * head.phase_dir[0] \
                            + rFOVz / 2 * head.slice_dir[0]
    corner_first_slice[1] = head.position[1] \
                            + rFOVx / 2 * head.read_dir[1] \
                            + rFOVy / 2 * head.phase_dir[1] \
                            + rFOVz / 2 * head.slice_dir[1]
    corner_first_slice[2] = head.position[2] \
                            + rFOVx / 2 * head.read_dir[2] \
                            + rFOVy / 2 * head.phase_dir[2] \
                            + rFOVz / 2 * head.slice_dir[2]
    h.update({"ImagePositionPatient": corner_first_slice})

    #Compute the corner position (ImagePositionPatient) of the last slice
    corner_last_slice[0] = head.position[0] \
                           + rFOVx / 2 * head.read_dir[0] \
                           + rFOVy / 2 * head.phase_dir[0] \
                           - rFOVz / 2 * head.slice_dir[0]
    corner_last_slice[1] = head.position[1] \
                           + rFOVx / 2 * head.read_dir[1] \
                           + rFOVy / 2 * head.phase_dir[1] \
                           - rFOVz / 2 * head.slice_dir[1]
    corner_last_slice[2] = head.position[2] \
                           + rFOVx / 2 * head.read_dir[2] \
                           + rFOVy / 2 * head.phase_dir[2] \
                           - rFOVz / 2 * head.slice_dir[2]

    h.update({"LastFile": {"ImagePositionPatient": corner_last_slice}})

    h.update({"Manufacturer": hdr.acquisitionSystemInformation.systemVendor})
    h.update({"Columns": hdr.encoding[0].reconSpace.matrixSize.y})
    h.update({"NiftiName": hdr.measurementInformation.protocolName})
    h.update({"InPlanePhaseEncodingDirection": 'COL'})

    return h